### PR TITLE
2D support for cuSTL::algorithm::mpi::Gather

### PR DESCRIPTION
**Pull request by Heikman**
2D support for cuSTL::algorithm::mpi::Gather

### CODE REVIEW

**Comment by ax3l** [comment]
can you pls directly add 1D support, too?

**Comment by Heikman** [comment]
just added 1D support

File: src/libPMacc/include/cuSTL/algorithm/mpi/Gather.tpp [hunk 1]
```diff
@@ -31,6 +31,41 @@ namespace algorithm
...
+namespace GatherHelper
+{
+    
+template<int dim, int memDim>
+struct posInMem;
+
+template<int dim>
+struct posInMem<dim, dim>
+{
+    math::Int<dim> operator()(const math::Int<dim>& pos, int) const
```
Comment (ax3l) [inline src/libPMacc/include/cuSTL/algorithm/mpi/Gather.tpp]: I am not sure that works for 4D and above ;) pls write explicit implementation for 1D, too and do not implement the general case, so it throws an error.

File: src/libPMacc/include/cuSTL/algorithm/mpi/Gather.tpp [hunk 2]
```diff
@@ -31,6 +31,41 @@ namespace algorithm
...
+    {
+        return pos;
+    }
+};
+
+template<>
+struct posInMem<3, 2>
+{
+    math::Int<2> operator()(const math::Int<3>& pos, int dir) const
```
Comment (ax3l) [inline src/libPMacc/include/cuSTL/algorithm/mpi/Gather.tpp]: pls use our defines, such as `DIM2`

**Comment by Heikman** [comment]
The explicit implementation of 1D is the general implementation. Note that even the "general implementation" is a specialization: ``dim == memDim``

**Comment by Heikman** [comment]
Everything's fine?

**Comment by ax3l** [comment]
ok with me, thanks!
@psychocoderHPC do you see room for improvement or can I merge? :)

**Comment by Heikman** [comment]
Here is psychocoderHPC, I hacked Heikman's account: yes you can merge.

**Comment by psychocoderHPC** [comment]
Nice try dude^^

**Comment by psychocoderHPC** [comment]
@ax3l  From my side all is ok

**Comment by ax3l** [comment]
;)

**Pull request merged**